### PR TITLE
dtlib: fix _raw_unit_addr() function

### LINKED ISSUE #77890: edtlib: wrong handling of "interrupt-map" property
**Describe the bug**

It's looks like some bug presented in the implementation of `edtlib._map_interrupt` function. As result `gen_defines.py` script is crashed on the property handling with a completely incomprehensible error, instead of generating headers.

**To Reproduce**

1. appy attached patch.
2. run west build -b ek_ra2l1
3. See error

**Expected behavior**

The `gen_defines.py`  should pass correctly and generate something resonable. Or generate a clear error message.

**Impact**

Any device tree with `interrupt-map` property.

**Logs and console output**

**Environment (please complete the following information):**

 - OS: any
 - Toolchain Zephyr SDK
 - Commit SHA or Version: 7b4e7ff9eda0227762666a39636b5f7729fdfd51

**Additional context**

[bad_interrupt_map.PATCH](https://github.com/user-attachments/files/16839060/bad_interrupt_map.PATCH)

### PR DESCRIPTION
**Pull request by avolkov-1221**
The '_raw_unit_addr()' function was incorrectly written a long time ago. It tries to look up and use the '#address-cells' of the node containing the 'interrupts' property itself, instead of using the '#address-cells' of the interrupt controller's node (or its parent).
 
 As a result, the mapping is successful only if the '#address-cells' value is the same in both the node and the interrupt controller, all other cases lead to random errors being generated by the 'gen_defines.py' script.
 
 Fixes https://github.com/zephyrproject-rtos/zephyr/issues/77890, https://github.com/zephyrproject-rtos/zephyr/issues/78020

### CODE REVIEW

**Comment by bjarki-andreasen** [comment]
Nice work! Could you extend the test suite https://github.com/zephyrproject-rtos/zephyr/tree/main/scripts/dts/python-devicetree/tests to catch this error? and for bonus points add a "practical" test to the devicetree test suite https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/lib/devicetree/api with a "fake" interrupt-map?

**Comment by avolkov-1221** [comment]
> Nice work! Could you extend the test suite https://github.com/zephyrproject-rtos/zephyr/tree/main/scripts/dts/python-devicetree/tests to catch this error? and for bonus points add a "practical" test to the devicetree test suite https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/lib/devicetree/api with a "fake" interrupt-map?

Ok. But as a following separate PR (I need to check the API a little bit).

**Comment by bjarki-andreasen** [comment]
Nice, tested it locally with your patch and this PR and `RA2_ICU_EVENT_SCI9_ERI` with prio 3 does indeed become IRQ 7 with prio 3.

from devicetree_generated.h
```
...
#define DT_N_S_soc_S_sci_40070120_P_interrupts_IDX_0 4122
#define DT_N_S_soc_S_sci_40070120_P_interrupts_IDX_1 3
...
#define DT_N_S_soc_S_sci_40070120_IRQ_IDX_0_VAL_irq 7
#define DT_N_S_soc_S_sci_40070120_IRQ_IDX_0_VAL_priority 3
```

**Review by bjarki-andreasen — APPROVED** [review]
Nice work!

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py [hunk 1]
```diff
@@ -3116,21 +3115,39 @@ def _pass_thru(
...
+    iparent: Optional[dtlib_Node] = parent
+    addr_len: Optional[int] = None
+
+    while iparent is not None:
+        addr_len = _address_cells_self(iparent)
+        if addr_len is None:
+            # check the parent(s) of interrupt-controller(s) too
+            iparent = iparent.parent
+        else:
+            break
```
Comment (57300) [inline scripts/dts/python-devicetree/src/devicetree/edtlib.py:3131]: This seems to treat `#address-cells` as inherited from a node's ancestor, which is contrary to what DTSpec says:
https://github.com/devicetree-org/devicetree-specification/blob/2697c66a59b9177f8ff7813c355cc7aa05f6e40a/source/chapter2-devicetree-basics.rst?plain=1#L644

> The *#address-cells* and *#size-cells* properties are not inherited from ancestors in the devicetree. They shall be explicitly defined.
>
> A DTSpec-compliant boot program shall supply *#address-cells* and *#size-cells* on all nodes that have children.
>
> If missing, a client program should assume a default value of 2 for *#address-cells*, and a value of 1 for *#size-cells*.

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py [hunk 2]
```diff
@@ -3116,21 +3115,39 @@ def _pass_thru(
...
         _err(f"{node!r} lacks 'reg' property "
              "(needed for 'interrupt-map' unit address lookup)")
 
-    addr_len = 4*_address_cells(node)
-
-    if len(node.props['reg'].value) < addr_len:
-        _err(f"{node!r} has too short 'reg' property "
+    addr_len *= 4
+    prop_len = len(node.props['reg'].value)
+    if  prop_len < addr_len or prop_len % 4 != 0:
```
Comment (57300) [inline scripts/dts/python-devicetree/src/devicetree/edtlib.py:3145]: Maybe also check that the `#address-cells` of `node.parent` is equal to that of `iparent`? I'm not sure what their exact relationship should be, but this `reg` property must definitely adhere to the `node.parent` one.

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py [hunk 3]
```diff
@@ -3116,21 +3115,39 @@ def _pass_thru(
...
+        _err(f"{node!r} has too short or incorrectly defined 'reg' property "
              "(while doing 'interrupt-map' unit address lookup)")
 
-    return node.props['reg'].value[:addr_len]
+    # address is a big endian value
+    return node.props['reg'].value[-addr_len:]
```
Comment (57300) [inline scripts/dts/python-devicetree/src/devicetree/edtlib.py:3150]: Is this correct? Taking `addr_len` from the end could mean taking size cells, or a different (address, size) pair.

I also see that this problem wouldn't be caught by tests, because the `interrupt-map-test` node has `#size-cells = <0>`.

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py [hunk 4]
```diff
@@ -3210,16 +3227,23 @@ def _phandle_val_list(
 
     return res
 
+def _address_cells_self(node: Optional[dtlib_Node]) -> Optional[int]:
```
Comment (57300) [inline scripts/dts/python-devicetree/src/devicetree/edtlib.py:3230]: nit: This could also replace or be used in `own_address_cells()`.

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py [hunk 5]
```diff
@@ -3210,16 +3227,23 @@ def _phandle_val_list(
...
+    # Returns the #address-cells setting for 'node', giving the number of <u32>
+    # cells used to encode the address in the 'reg' property
```
Comment (57300) [inline scripts/dts/python-devicetree/src/devicetree/edtlib.py:3232]: This comment is the same as for `_address_cells()`. Maybe clarify how this new function differs? (own property vs. parent bus property)

Comment (avolkov-1221) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py:3232]: Actually the comment should be fixed for `_address_cells()`, since it returns the parent's `#address-cells` :). But the comment for `own_address_cells()` is correct: it returns the value for the node, or `None` if it doesn't have one.

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py (hunk 4 above)
Comment (avolkov-1221) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py:3230]: Ok. Accepted. Looks better.

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py (hunk 3 above)
Comment (avolkov-1221) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py:3150]: Ah, yes, thank you. I forgot about the size. Will fix it.

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py (hunk 2 above)
Comment (avolkov-1221) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py:3145]: I'll reread the specs more carefully.

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py (hunk 1 above)
Comment (avolkov-1221) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py:3131]: Thank you for the reference. I'll rework the algo.

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py (hunk 2 above)
Comment (avolkov-1221) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py:3145]: @57300 @bjarki-andreasen @nashif  @galak  It's looks like we have yet another DT interrupts related problem: `_interrupt_parent()` implementation.

The current implementation is scanning the tree for the 'interrupt-parent' until root node. But the latest DT specification is clear:
> Because the hierarchy of the nodes in the interrupt tree might not match the devicetree, the interrupt-parent property is available to make the definition of an interrupt parent explicit. The value is the phandle to the interrupt parent. **If this property is missing from a device, its interrupt parent is assumed to be its devicetree parent.**

And the latest Linux kernel already follow the above spec. Check the [vexpress-v2m.dtsi](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/arch/arm/boot/dts/arm/vexpress-v2m.dtsi), as an example. Obviously I should modify the `_interrupt_parent()` in this PR too, because the correct `interrupt-map` functionality  is depends on it, but consequences are not clear for me now. Since the `interrupt-map` property is actually used only for one board, I can make almost any modification for it. However, `_interrupt_parent()` function is an another beast: it's part of the core DT functionality, so a simple modification can break many targets. Should I put 'Architecture Review' on it (it's looks like "treewide" modif now)?

Comment (bjarki-andreasen) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py:3145]: To clarify, the issue is that we are walking up the tree until finding an interrupt controller, rather than asserting that the parent node is the interrupt controller, and generating a fault if this is not the case?

Comment (avolkov-1221) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py:3145]: > To clarify, the issue is that we are walking up the tree until finding an interrupt controller, rather than asserting that the parent node is the interrupt controller, and generating a fault if this is not the case?

A little bit different: the `_interrupt_parent()` should scan only and only the parent node, and use the `interrupt-parent` property, if it there, or use the parent node, if not. No matter if the parent node have or have not `interrupt-controller` property. In this case the ICU can have `interrupt-controller` (as it's more logical), and the `soc` become `interrupt-map` node (also using the address-cell and size from the `soc` becomes more agreeable). Check the 'vexpress-v2m.dtsi' file mentioned above, where interrupt-map is used instead of `interrupt-parent` (which would be clearer), but not instead of `interrupt-controller`.
However this new behavior of  `_interrupt_parent()` (and trivial modification), can break multiple targets. Or not. I haven't checked yet.

Comment (57300) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py:3145]: I think we should be careful here. Firstly, the part of the spec that @avolkov-1221 quoted isn't new; git blame shows that the wording hasn't changed in 8 years: https://github.com/devicetree-org/devicetree-specification/blame/2697c66a59b9177f8ff7813c355cc7aa05f6e40a/source/chapter2-devicetree-basics.rst#L1076-L1080

But setting aside the spec for a moment, we should also take a look at the reference implementation - dtc, which is more actively maintained than the spec, as far as I'm aware. dtc will produce warnings on incorrect use of interrupt properties, and one of the checks is written here: https://github.com/dgibson/dtc/blob/99031e3a4a6e479466ae795790b44727434ca27d/checks.c#L1686

Here we can see that finding an interrupt parent does involve walking up the tree. The pseudocode would go something like this (written recursively):
```
proc get_interrupt_parent(node):
    if "interrupt-parent" in node.props:
        phandle := node.props["interrupt-parent"]
        iparent := get_node(phandle)
        assert iparent is interrupt controller or nexus
        return iparent

    if node is root:
        error "no interrupt parent"

    if node.parent is interrupt controller or nexus:
        return node.parent

    return get_interrupt_parent(node.parent)
```

This would suggest that the edtlib implementation is partially correct, only missing two things:
1. If an ancestor has an `interrupt-controller` or `interrupt-map` property, the walk must terminate.
2. If an `interrupt-parent` property is found, the linked node must be a valid interrupt controller or nexus.

I think it's strange that the spec seems to contradict this, by suggesting that only a node's immediate parent - not just any ancestor - can determine its interrupt parent. Even `vexpress-v2m.dtsi` doesn't abide by that rule - see how every interrupt-generating node (e.g. `/bus@40000000/motherboard-bus@40000000/iofpga@7,00000000/clcd@1f000`) is more than one level below the interrupt nexus (`/bus@40000000`).

On the other hand, I've found the spec to be incomplete and imprecise in several spots before. In fact, this very PR fixes a bug which likely stems from the spec's `interrupt-map` definition being misleading:
> **child unit address**
> The unit address of the child node being mapped. The number of 32-bit cells required to specify this is described by the *#address-cells* property of the **bus node on which the child is located**.

In conclusion, I think the best choice would be to align edtlib with dtc, which would also minimize the breakage. Regardless, I also think this can be taken in a separate PR, as the `#address-cells` fix is valuable enough on its own.

Comment (bjarki-andreasen) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py:3145]: nice research! @57300 :)

Comment (avolkov-1221) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py:3145]: > I think we should be careful here. Firstly, the part of the spec that @avolkov-1221 quoted isn't new; git blame shows that the wording hasn't changed in 8 years: https://github.com/devicetree-org/devicetree-specification/blame/2697c66a59b9177f8ff7813c355cc7aa05f6e40a/source/chapter2-devicetree-basics.rst#L1076-L1080

The 'new' here means newer than this comment in `_interrupt_parent()` :) :
```
 # Returns the node pointed at by the closest 'interrupt-parent', searching
 # the parents of 'node'. As of writing, this behavior isn't specified in
 # the DT spec., but seems to match what some .dts files except.
```
As you pointed out above, the behavior is specified in the spec for years, or the commenter meant something else.

>  
>  
... Snip ...

>This would suggest that the edtlib implementation is partially correct, only missing two things:
>
>If an ancestor has an interrupt-controller or interrupt-map property, the walk must terminate.
>If an interrupt-parent property is found, the linked node must be a valid interrupt controller or nexus.
>I think it's strange that the spec seems to contradict this, by suggesting that only a node's immediate parent - not just any ancestor - can determine its interrupt parent. Even vexpress-v2m.dtsi doesn't abide by that rule - see how every interrupt-generating node (e.g. /bus@40000000/motherboard-bus@40000000/iofpga@7,00000000/clcd@1f000) is more than one level below the interrupt nexus (/bus@40000000).

I suppose authors of the specification is mainly concerned with defining software behavior and seems to consider error checking outside its scope. Btw the `interrupt-parent` is propagated to all its children, isn't it? Also, dtc treats the `interrupt-map` the same way as `interrupt-controller` and `interrupt-parent`, i.e. as an inheritable property (I don't know where this behavior comes from).

... Snip ...

> 
> In conclusion, I think the best choice would be to align edtlib with dtc, which would also minimize the breakage. Regardless, I also think this can be taken in a separate PR, as the `#address-cells` fix is valuable enough on its own.

Agree that align `_interrupt_parent()` with dtc is the best solution for now, however I'm not sure if it would be reasonable to make this as a separate PR rather than a different commit in this one (actually it's simple, a "couple lines" modification). After updating `_raw_unit_addr()` et al, creating the nexus as done in vexpress-v2m.dtsi will still be impossible. I mean this part of the code:
```
	bus@40000000 {
		compatible = "simple-bus";
		#address-cells = <1>;
		#size-cells = <1>;
		....
		#interrupt-cells = <1>;
		interrupt-map-mask = <0 63>;
		interrupt-map = <0  0 &gic GIC_SPI  0 IRQ_TYPE_LEVEL_HIGH>,
		....
```
The current `_interrupt_parent()` skips such declarations (regardless of whether the `interrupt-parent` property is present or missing). Additionally, the intermediate interrupt controller will still need to be defined somewhat out of the spec: without `interrupt-controller`, but with the `interrupt-map` property. Probably it will be better to add yet another issue with ref to this PR.

**Comment by soburi** [comment]
It looked like the review request hadn't been resubmitted and was stuck, so I made a new request. 

@57300 @avolkov-1221, thank you in advance.

**Comment by 57300** [comment]
> It looked like the review request hadn't been resubmitted and was stuck, so I made a new request.
> 
> @57300 @avolkov-1221, thank you in advance.

Thanks. I see that this PR is yet to be updated following our last discussion, so I'm looking forward to that.

**Comment by kartben** [comment]
@avolkov-1221 will you be able to revisit this PR?
